runtime/runtime_impl_jetty9/src/main/java/com/google/apphosting/runtime/jetty9/JettyHttpHandler.java

```diff
@@ -137,8 +139,21 @@ public void handle(
       // We will report the exception via the rpc. We don't mark this thread as interrupted because
       // ThreadGroupPool would use that as a signal to remove the thread from the pool; we don't
       // need that.
-      handleException(ex, requestToken, genericResponse);
-      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, ex.getMessage());
+      Throwable cause = unwrap(ex, BadMessageException.class, UnavailableException.class);
+      handleException(cause, requestToken, genericResponse);
+      if (cause instanceof BadMessageException) {
+        BadMessageException bme = (BadMessageException) cause;
+        response.sendError(bme.getCode(), cause.getMessage());
+      } else if (cause instanceof UnavailableException) {
+        UnavailableException ue = (UnavailableException) cause;
+        response.sendError(
+            ue.isPermanent()
+                ? HttpServletResponse.SC_NOT_FOUND
+                : HttpServletResponse.SC_SERVICE_UNAVAILABLE,
+            cause.getMessage());
+      } else {
```

Comment:
Curious on 
1. What response code do we give in Jetty 12? Also the RPC Connector mode?
2. I see that we are returning 404 but the PR mentions 413

Reply:
The `UnavailableException` has nothing to do with the `SizeLimitHandlerTest`, only the `BadMessageException` does. Previously in this method we were just sending 500 response, but the response really depends on the exception, so in case of a `BadMessageException` we should send a 400 and with a `UnavailableException` we send either 404 or 503 depending if its permanently unavailable.

The Jetty 12 runtime does the same thing just through a different mechanism, it will not even reach this point, it handles it internal to the `ServletChannel` and writes the error itself with the correct code. In the case that it did reach this point in the Jetty 12 runtime (by failing the callback), then it uses `Response.writeError` which will deal with `HttpException` to send a correct response code.